webapp/src/js/services/contact-view-model-generator.js

```diff
@@ -67,146 +100,134 @@ angular.module('inboxServices').factory('ContactViewModelGenerator',
-        .then(function() {
-          return children;
-        });
+    const sortChildren = function(model, childModels) {
+      childModels.forEach(group => {
+        const comparator = group.type && group.type.sort_by_dob ? AGE_COMPARATOR : NAME_COMPARATOR;
+        group.contacts.sort(comparator);
+      });
+      childModels.sort(TYPE_COMPARATOR);
+      return childModels;
     };
 
-    var sortChildren = function(model, children) {
-      if (children.places) {
-        children.places.sort(NAME_COMPARATOR);
+    const getPersonChildTypes = (types, parentId) => {
+      if (!parentId) {
+        return [];
       }
-      if (children.persons) {
-        var personComparator = model.doc.type === 'clinic' ? AGE_COMPARATOR : NAME_COMPARATOR;
-        children.persons.sort(personComparator);
-      }
-      return sortPrimaryContactToTop(model, children);
+      const childTypes = types.filter(type => {
+        return type.person &&
+               type.parents &&
+               type.parents.includes(parentId);
+      });
+      return childTypes;
     };
 
-    const getChildren = (contactId, { getChildPlaces } = {}) => {
+    const getChildren = (model, types, { getChildPlaces } = {}) => {
       const options = { include_docs: true };
+      const contactId = model.doc._id;
       if (getChildPlaces) {
         // get all types
         options.startkey = [ contactId ];
         options.endkey = [ contactId, {} ];
       } else {
-        // just get people
-        options.key = [ contactId, 'person' ];
+        // just get person children
+        const childTypes = getPersonChildTypes(types, model.type && model.type.id);
+        if (!childTypes.length) {
+          return $q.resolve([]);
+        }
+        options.keys = childTypes.map(type => [ contactId, type.id ]);
       }
       return DB().query('medic-client/contacts_by_parent', options)
         .then(response => response.rows);
     };
 
-    var loadChildren = function(model, options) {
-      model.children = {};
-      if (model.doc.type === 'person') {
-        return $q.resolve({});
-      }
+    const buildChildModels = (groups, types) => {
+      return Object.keys(groups).map(typeId => {
+        return {
+          contacts: groups[typeId],
+          type: types.find(type => type.id === typeId)
+        };
+      });
+    };
 
-      return getChildren(model.doc._id, options)
-        .then(splitContactsByType)
-        .then(function(children) {
-          if (children.places && children.places.length) {
-            var childPlacesSchema = ContactSchema.get(children.places[0].doc.type);
-            children.childPlacesLabel = childPlacesSchema.pluralLabel;
-            children.childPlacesIcon = childPlacesSchema.icon;
-          }
-          return sortChildren(model, children);
-        });
+    const markDeceased = (model, children) => {
+      children.forEach(child => {
+        if (child.doc.date_of_death) {
+          model.deceasedCount++;
```

Comment:
You're updating the model here, but then the Redux reducer only updates `model.children`, so your `model.deceasedCount` is lost and will never be set. Maybe it's simpler to `markDeceased` after grouping and save `deceasedCount` per group?

Reply:
Oh, for sure! And storing it outside the groups is a bug anyway because the template loops over the groups and decides whether to show the "deceased" footer based on that shared deceasedCount... I'm on it!